File: backend/services/workflow-engine-go-service/pkg/clients/safety_gateway_client.go
package clients

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"go.uber.org/zap"
)

// SafetyGatewayClient defines the interface for Safety Gateway communication
type SafetyGatewayClient interface {
	ComprehensiveValidation(ctx context.Context, request *SafetyValidationRequest) (*SafetyValidationResponse, error)
	ValidateProposal(ctx context.Context, request *ProposalValidationRequest) (*ValidationResult, error)
	HealthCheck(ctx context.Context) error
}

// SafetyValidationRequest represents a comprehensive validation request
type SafetyValidationRequest struct {
	ProposalSetID           string                   `json:"proposal_set_id"`
	SnapshotID              string                   `json:"snapshot_id"`
	Proposals               []map[string]interface{} `json:"proposals"`
	PatientContext          map[string]interface{}   `json:"patient_context"`
	ValidationRequirements  map[string]interface{}   `json:"validation_requirements"`
	CorrelationID           string                   `json:"correlation_id"`
	ValidationScope         []string                 `json:"validation_scope,omitempty"`
	RiskTolerance          string                   `json:"risk_tolerance,omitempty"`
}

// ProposalValidationRequest represents a single proposal validation
type ProposalValidationRequest struct {
	ProposalID      string                 `json:"proposal_id"`
	SnapshotID      string                 `json:"snapshot_id"`
	Proposal        map[string]interface{} `json:"proposal"`
	PatientContext  map[string]interface{} `json:"patient_context"`
	ValidationLevel string                 `json:"validation_level"` // "basic", "comprehensive", "critical"
	CorrelationID   string                 `json:"correlation_id"`
}

// SafetyValidationResponse represents the comprehensive validation response
type SafetyValidationResponse struct {
	ValidationID         string                   `json:"validation_id"`
	Verdict              string                   `json:"verdict"` // "SAFE", "WARNING", "UNSAFE", "ERROR"
	OverallRiskScore     float64                  `json:"overall_risk_score"`
	Findings             []ValidationFinding      `json:"findings"`
	OverrideTokens       []string                 `json:"override_tokens,omitempty"`
	OverrideRequirements map[string]interface{}   `json:"override_requirements,omitempty"`
	ValidationMetrics    map[string]interface{}   `json:"validation_metrics"`
	ExecutedEngines      []string                 `json:"executed_engines"`
	Status               string                   `json:"status"`
	Message              string                   `json:"message,omitempty"`
}

// ValidationFinding represents a single validation finding
type ValidationFinding struct {
	FindingID            string                 `json:"finding_id"`
	Severity             string                 `json:"severity"` // "LOW", "MEDIUM", "HIGH", "CRITICAL"
	Category             string                 `json:"category"`
	Description          string                 `json:"description"`
	ClinicalSignificance string                 `json:"clinical_significance"`
	Recommendation       string                 `json:"recommendation"`
	ConfidenceScore      float64                `json:"confidence_score"`
	Source               string                 `json:"source"`
	Evidence             map[string]interface{} `json:"evidence,omitempty"`
	Overridable          bool                   `json:"overridable"`
}

// ValidationResult represents a validation result
type ValidationResult struct {
	ValidationID    string              `json:"validation_id"`
	ProposalID      string              `json:"proposal_id"`
	Verdict         string              `json:"verdict"`
	RiskScore       float64             `json:"risk_score"`
	Findings        []ValidationFinding `json:"findings"`
	ProcessingTime  float64             `json:"processing_time_ms"`
	ValidatedBy     []string            `json:"validated_by"`
}

// safetyGatewayClientImpl implements SafetyGatewayClient
type safetyGatewayClientImpl struct {
	baseURL    string
	httpClient *http.Client
	logger     *zap.Logger
}

// NewSafetyGatewayClient creates a new Safety Gateway client
func NewSafetyGatewayClient(baseURL string, logger *zap.Logger) SafetyGatewayClient {
	return &safetyGatewayClientImpl{
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: 45 * time.Second, // Longer timeout for comprehensive validation
			Transport: &http.Transport{
				MaxIdleConns:        10,
				MaxIdleConnsPerHost: 10,
				IdleConnTimeout:     60 * time.Second,
			},
		},
		logger: logger,
	}
}

// ComprehensiveValidation performs comprehensive safety validation
func (c *safetyGatewayClientImpl) ComprehensiveValidation(ctx context.Context, request *SafetyValidationRequest) (*SafetyValidationResponse, error) {
	c.logger.Info("Starting comprehensive safety validation",
		zap.String("correlation_id", request.CorrelationID),
		zap.String("snapshot_id", request.SnapshotID),
		zap.Int("proposal_count", len(request.Proposals)))

	// Set default validation requirements if not specified
	if request.ValidationRequirements == nil {
		request.ValidationRequirements = map[string]interface{}{
			"cae_engine":              true,
			"protocol_engine":         true,
			"interaction_engine":      true,
			"contraindication_engine": true,
			"allergy_engine":          true,
			"dosing_engine":           true,
			"comprehensive_validation": true,
		}
	}

	// Prepare request body
	requestBody, err := json.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create HTTP request
	url := fmt.Sprintf("%s/api/v1/validation/comprehensive", c.baseURL)
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(requestBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("User-Agent", "workflow-engine-service/1.0")
	httpReq.Header.Set("X-Correlation-ID", request.CorrelationID)
	httpReq.Header.Set("X-Validation-Priority", "HIGH") // Priority for workflow orchestration

	// Execute request
	startTime := time.Now()
	resp, err := c.httpClient.Do(httpReq)
	duration := time.Since(startTime)

	c.logger.Info("Safety Gateway validation completed",
		zap.String("correlation_id", request.CorrelationID),
		zap.Duration("duration", duration),
		zap.Int("status_code", func() int {
			if resp != nil {
				return resp.StatusCode
			}
			return 0
		}()))

	if err != nil {
		c.logger.Error("Failed to call Safety Gateway",
			zap.String("correlation_id", request.CorrelationID),
			zap.Error(err))
		return nil, fmt.Errorf("failed to execute validation request: %w", err)
	}
	defer resp.Body.Close()

	// Check response status
	if resp.StatusCode != http.StatusOK {
		c.logger.Error("Safety Gateway returned error",
			zap.String("correlation_id", request.CorrelationID),
			zap.Int("status_code", resp.StatusCode))
		return nil, fmt.Errorf("Safety Gateway returned status %d", resp.StatusCode)
	}

	// Parse response
	var response SafetyValidationResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	// Validate response
	if response.ValidationID == "" {
		return nil, fmt.Errorf("invalid response: missing validation_id")
	}

	// Log validation results
	c.logger.Info("Safety validation completed",
		zap.String("correlation_id", request.CorrelationID),
		zap.String("validation_id", response.ValidationID),
		zap.String("verdict", response.Verdict),
		zap.Float64("risk_score", response.OverallRiskScore),
		zap.Int("findings_count", len(response.Findings)),
		zap.Strings("engines", response.ExecutedEngines))

	// Log critical findings
	for _, finding := range response.Findings {
		if finding.Severity == "CRITICAL" || finding.Severity == "HIGH" {
			c.logger.Warn("Critical safety finding",
				zap.String("correlation_id", request.CorrelationID),
				zap.String("finding_id", finding.FindingID),
				zap.String("severity", finding.Severity),
				zap.String("category", finding.Category),
				zap.String("description", finding.Description))
		}
	}

	return &response, nil
}

// ValidateProposal validates a single proposal
func (c *safetyGatewayClientImpl) ValidateProposal(ctx context.Context, request *ProposalValidationRequest) (*ValidationResult, error) {
	c.logger.Info("Validating single proposal",
		zap.String("correlation_id", request.CorrelationID),
		zap.String("proposal_id", request.ProposalID),
		zap.String("validation_level", request.ValidationLevel))

	// Prepare request body
	requestBody, err := json.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create HTTP request
	url := fmt.Sprintf("%s/api/v1/validation/proposal", c.baseURL)
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(requestBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("User-Agent", "workflow-engine-service/1.0")
	httpReq.Header.Set("X-Correlation-ID", request.CorrelationID)

	// Execute request
	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		c.logger.Error("Failed to validate proposal",
			zap.String("correlation_id", request.CorrelationID),
			zap.String("proposal_id", request.ProposalID),
			zap.Error(err))
		return nil, fmt.Errorf("failed to execute validation request: %w", err)
	}
	defer resp.Body.Close()

	// Check response status
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("Safety Gateway returned status %d", resp.StatusCode)
	}

	// Parse response
	var response ValidationResult
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	c.logger.Info("Proposal validation completed",
		zap.String("correlation_id", request.CorrelationID),
		zap.String("proposal_id", request.ProposalID),
		zap.String("validation_id", response.ValidationID),
		zap.String("verdict", response.Verdict),
		zap.Float64("risk_score", response.RiskScore),
		zap.Float64("processing_time", response.ProcessingTime))

	return &response, nil
}

// HealthCheck checks if Safety Gateway is healthy
func (c *safetyGatewayClientImpl) HealthCheck(ctx context.Context) error {
	url := fmt.Sprintf("%s/health", c.baseURL)
	
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return fmt.Errorf("failed to create health check request: %w", err)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return fmt.Errorf("health check failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("Safety Gateway unhealthy: status %d", resp.StatusCode)
	}

	// Parse health response to check validation engines status
	var healthResponse struct {
		Status  string            `json:"status"`
		Engines map[string]string `json:"engines,omitempty"`
	}
	
	if err := json.NewDecoder(resp.Body).Decode(&healthResponse); err != nil {
		c.logger.Warn("Failed to decode health response", zap.Error(err))
		// Still return nil if basic health check passed
		return nil
	}

	// Check individual engines if available
	if healthResponse.Engines != nil {
		for engine, status := range healthResponse.Engines {
			if status != "healthy" && status != "ok" {
				c.logger.Warn("Safety Gateway engine unhealthy",
					zap.String("engine", engine),
					zap.String("status", status))
			}
		}
	}

	return nil
}

File: backend/services/workflow-engine-service/app/integration/safety_gateway_client.py
"""
Safety Gateway Client for Strategic Orchestration

This module provides integration with the Safety Gateway Platform
for the VALIDATE step of the Calculate > Validate > Commit pattern.
"""

import logging
import asyncio
from typing import Dict, List, Any, Optional
from dataclasses import dataclass
from enum import Enum
import httpx
import json
from datetime import datetime

logger = logging.getLogger(__name__)


class ValidationVerdict(Enum):
    """Safety validation verdict enumeration"""
    SAFE = "SAFE"
    WARNING = "WARNING"  
    UNSAFE = "UNSAFE"
    ERROR = "ERROR"


class FindingSeverity(Enum):
    """Clinical finding severity levels"""
    CRITICAL = "CRITICAL"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"
    INFO = "INFO"


@dataclass
class ValidationFinding:
    """Individual validation finding"""
    finding_id: str
    severity: FindingSeverity
    category: str  # DDI, ALLERGY, DOSE, CONTRAINDICATION, etc.
    description: str
    clinical_significance: str
    recommendation: str
    source_rule: str
    confidence_score: float


@dataclass
class SafetyValidationRequest:
    """Request structure for Safety Gateway validation"""
    proposal_set_id: str
    snapshot_id: str  # Same snapshot used in Calculate step
    proposals: List[Dict[str, Any]]
    patient_context: Dict[str, Any]
    validation_requirements: Dict[str, Any]
    correlation_id: str
    urgency: str = "ROUTINE"


@dataclass 
class SafetyValidationResponse:
    """Response structure from Safety Gateway validation"""
    validation_id: str
    verdict: ValidationVerdict
    overall_risk_score: float
    findings: List[ValidationFinding]
    validation_summary: Dict[str, Any]
    
    # Override capabilities
    override_tokens: Optional[List[str]] = None
    override_requirements: Optional[Dict[str, Any]] = None
    
    # Timing and metadata
    validation_time_ms: float = 0.0
    engines_used: List[str] = None
    kb_versions: Dict[str, str] = None


class SafetyGatewayClient:
    """
    Client for Safety Gateway Platform integration
    
    Coordinates comprehensive safety validation using:
    - CAE (Clinical Alert Engine) 
    - Protocol Engine
    - Drug-drug interaction screening
    - Allergy checking
    - Dose validation
    - Contraindication screening
    """
    
    def __init__(self, gateway_url: str = "http://localhost:8018"):
        self.gateway_url = gateway_url
        self.http_client = httpx.AsyncClient(timeout=30.0)
        
        # Validation configuration
        self.validation_config = {
            "enable_cae_engine": True,
            "enable_protocol_engine": True,
            "enable_ddi_screening": True,
            "enable_allergy_checking": True,
            "enable_dose_validation": True,
            "enable_contraindication_screening": True,
            "parallel_validation": True,  # Run engines in parallel
            "fail_fast": False,  # Complete all validations even if one fails
            "include_warnings": True,
            "generate_override_tokens": True
        }
    
    async def comprehensive_validation(
        self, 
        request: SafetyValidationRequest
    ) -> SafetyValidationResponse:
        """
        Perform comprehensive safety validation via Safety Gateway
        
        This is the VALIDATE step in Calculate > Validate > Commit pattern.
        Uses the same snapshot ID to ensure data consistency with Calculate step.
        """
        validation_start = datetime.utcnow()
        
        logger.info(f"Starting comprehensive validation {request.correlation_id} for snapshot {request.snapshot_id}")
        
        try:
            # Prepare validation request for Safety Gateway
            gateway_request = {
                "validation_type": "COMPREHENSIVE_MEDICATION_SAFETY",
                "request_id": request.correlation_id,
                "snapshot_id": request.snapshot_id,  # Critical for consistency
                "proposal_set_id": request.proposal_set_id,
                "proposals": request.proposals,
                "patient_context": request.patient_context,
                "validation_config": {
                    **self.validation_config,
                    **request.validation_requirements
                },
                "urgency": request.urgency,
                "timestamp": validation_start.isoformat()
            }
            
            # Call Safety Gateway comprehensive validation endpoint
            response = await self.http_client.post(
                f"{self.gateway_url}/api/v1/safety/comprehensive-validation",
                json=gateway_request,
                headers={
                    "Content-Type": "application/json",
                    "X-Correlation-ID": request.correlation_id,
                    "X-Validation-Source": "workflow-platform"
                }
            )
            response.raise_for_status()
            
            gateway_result = response.json()
            validation_time = (datetime.utcnow() - validation_start).total_seconds() * 1000
            
            # Parse findings from Safety Gateway response
            findings = []
            for finding_data in gateway_result.get("findings", []):
                finding = ValidationFinding(
                    finding_id=finding_data["finding_id"],
                    severity=FindingSeverity(finding_data["severity"]),
                    category=finding_data["category"],
                    description=finding_data["description"],
                    clinical_significance=finding_data["clinical_significance"],
                    recommendation=finding_data["recommendation"],
                    source_rule=finding_data["source_rule"],
                    confidence_score=finding_data["confidence_score"]
                )
                findings.append(finding)
            
            # Determine overall verdict based on findings
            verdict = self._determine_verdict(findings, gateway_result)
            
            return SafetyValidationResponse(
                validation_id=gateway_result["validation_id"],
                verdict=verdict,
                overall_risk_score=gateway_result["overall_risk_score"],
                findings=findings,
                validation_summary=gateway_result["validation_summary"],
                override_tokens=gateway_result.get("override_tokens"),
                override_requirements=gateway_result.get("override_requirements"),
                validation_time_ms=validation_time,
                engines_used=gateway_result.get("engines_used", []),
                kb_versions=gateway_result.get("kb_versions", {})
            )
            
        except httpx.HTTPStatusError as e:
            logger.error(f"Safety Gateway HTTP error {e.response.status_code}: {e.response.text}")
            return self._create_error_response(request.correlation_id, f"HTTP {e.response.status_code}: {e.response.text}")
            
        except Exception as e:
            logger.error(f"Safety Gateway validation failed for {request.correlation_id}: {str(e)}")
            return self._create_error_response(request.correlation_id, str(e))
    
    async def validate_override_request(
        self,
        validation_id: str,
        override_tokens: List[str],
        provider_justification: str,
        correlation_id: str
    ) -> Dict[str, Any]:
        """
        Validate provider override request
        
        Ensures override tokens are valid and provider has appropriate privileges.
        """
        try:
            override_request = {
                "validation_id": validation_id,
                "override_tokens": override_tokens,
                "provider_justification": provider_justification,
                "correlation_id": correlation_id,
                "timestamp": datetime.utcnow().isoformat()
            }
            
            response = await self.http_client.post(
                f"{self.gateway_url}/api/v1/safety/validate-override",
                json=override_request,
                headers={
                    "Content-Type": "application/json",
                    "X-Correlation-ID": correlation_id
                }
            )
            response.raise_for_status()
            
            return response.json()
            
        except Exception as e:
            logger.error(f"Override validation failed for {correlation_id}: {str(e)}")
            return {
                "status": "OVERRIDE_VALIDATION_FAILED",
                "error": str(e)
            }
    
    async def get_validation_status(self, validation_id: str) -> Dict[str, Any]:
        """Get status of a validation request"""
        try:
            response = await self.http_client.get(
                f"{self.gateway_url}/api/v1/safety/validation/{validation_id}/status"
            )
            response.raise_for_status()
            
            return response.json()
            
        except Exception as e:
            logger.error(f"Failed to get validation status for {validation_id}: {str(e)}")
            return {
                "status": "STATUS_UNAVAILABLE",
                "error": str(e)
            }
    
    async def health_check(self) -> Dict[str, Any]:
        """Check Safety Gateway health"""
        try:
            response = await self.http_client.get(f"{self.gateway_url}/health")
            
            if response.status_code == 200:
                health_data = response.json()
                return {
                    "status": "healthy",
                    "safety_gateway": health_data,
                    "validation_capabilities": [
                        "CAE Engine",
                        "Protocol Engine", 
                        "DDI Screening",
                        "Allergy Checking",
                        "Dose Validation",
                        "Contraindication Screening"
                    ]
                }
            else:
                return {
                    "status": "unhealthy",
                    "error": f"HTTP {response.status_code}"
                }
                
        except Exception as e:
            return {
                "status": "unavailable",
                "error": str(e)
            }
    
    def _determine_verdict(
        self, 
        findings: List[ValidationFinding], 
        gateway_result: Dict[str, Any]
    ) -> ValidationVerdict:
        """Determine overall validation verdict based on findings"""
        
        # Check for critical or high severity findings
        critical_findings = [f for f in findings if f.severity == FindingSeverity.CRITICAL]
        high_findings = [f for f in findings if f.severity == FindingSeverity.HIGH]
        
        if critical_findings:
            return ValidationVerdict.UNSAFE
        elif high_findings:
            return ValidationVerdict.WARNING
        elif gateway_result.get("overall_risk_score", 0) > 0.7:
            return ValidationVerdict.WARNING
        else:
            return ValidationVerdict.SAFE
    
    def _create_error_response(
        self, 
        correlation_id: str, 
        error_message: str
    ) -> SafetyValidationResponse:
        """Create error response for validation failures"""
        return SafetyValidationResponse(
            validation_id="",
            verdict=ValidationVerdict.ERROR,
            overall_risk_score=1.0,  # Max risk for errors
            findings=[
                ValidationFinding(
                    finding_id="VALIDATION_ERROR",
                    severity=FindingSeverity.CRITICAL,
                    category="SYSTEM_ERROR",
                    description=f"Safety validation system error: {error_message}",
                    clinical_significance="Cannot determine medication safety",
                    recommendation="Manual clinical review required",
                    source_rule="SYSTEM_ERROR_HANDLER",
                    confidence_score=1.0
                )
            ],
            validation_summary={
                "error": error_message,
                "status": "VALIDATION_FAILED"
            },
            validation_time_ms=0.0,
            engines_used=[],
            kb_versions={}
        )
    
    async def close(self):
        """Close HTTP client"""
        await self.http_client.aclose()


# Global client instance
safety_gateway_client = SafetyGatewayClient()